### DISCUSSION TITLE
Configurable Antialiasing

### DISCUSSION THREAD

**Discussion by DennisGruendig** [Chlumsky/msdfgen#194] [Q&A] [answered]
Hi there, I was using the shader to render some fonts and I really love the overall look, except for one thing: pixelated fonts.

I was wondering if you could tell of me a clean way to change the antialiasing behaviour, as I am curently unable to fully understand the entirety of the distance calculations.

The most ideal way would be for a float value, indicating the "smoothness". But which of the values would I need to multiply with that value, in order so make the edge aliased.

I figured out that if you change the unitRange to 1, it will end up aliased but this is surely not the right way, is it? Also the dot product is what gets me confused.

![grafik](https://github.com/Chlumsky/msdfgen/assets/86776302/afbd66d5-d6da-4ed1-ba2d-2d04994aee1e)

**Comment by Chlumsky** [comment, marked as answer]
I don't understand what exactly you mean by "pixelated fonts", "make the edge aliased". Post an example of what you want it to look like.

If you want to disable anti-aliasing and just make each pixel fully transparent or opaque (that's what I might call "pixelated"), then do:
```hlsl
float sd = median(msd.r, msd.g, msd.b);
float opacity = step(0.5, sd);
return lerp(float4(0, 0, 0, 0), fgColor, opacity);
```

**Reply by DennisGruendig** [reply]
This is not the answer to my problem but definitely helpful, thank you!

I think I have more or less accomplished what I wanted now. I can adjust the sharpness by allowing different levels of opacity, making it at least seem more sharp when the value is pushed up. This makes the edge kinda distorted though, now that a think of it. Maybe I should push the value down instead of up... but it's a start.

**This is my shader now:**
```hlsl
float softOpacity = clamp(screenPxDistance + 0.5, 0.0, 1.0);
float sharpOpacity = step(0.5, sd);
float opacity = lerp(softOpacity, sharpOpacity, smoothness);

return lerp(0, fgColor, opacity);
```

**And these are the resulting graphics:**
![Screenshot 2024-01-15 155356](https://github.com/Chlumsky/msdfgen/assets/86776302/f0fb9d95-ec82-4cae-a419-897c84f3b3f1)
![Screenshot 2024-01-15 155346](https://github.com/Chlumsky/msdfgen/assets/86776302/a505db95-3f26-4cc4-b75b-9564fed3bb5e)
![Screenshot 2024-01-15 155324](https://github.com/Chlumsky/msdfgen/assets/86776302/f4ae8775-2956-4f31-8df1-ffce0649737a)

**Reply by Chlumsky** [reply]
I have zero idea of what you're trying to accomplish.

**Reply by DennisGruendig** [reply]
I‘m trying to get pixelated fonts to render without smoothing, as they are meant to be pixelated. While the antialiasing is perfect for regular fonts, it really doesn‘t look right on the pixel fonts.

The font in the pictures might not have been the right choice to demonstrate the problem, but they clearly show the rendering difference.

I just wanted sharp edged, thats all…

**Reply by Chlumsky** [reply]
I still don't know what you mean. Pixelated fonts to me means raster fonts, which is a completely separate thing from vector-based fonts, which MSDFgen uses. Your last picture seems pretty sharp edged to me.

**Reply by DennisGruendig** [reply]
When using pixelated fonts (not bitmap fonts), based on vectors, which are not perfectly aligned to a pixel grid

![Screenshot 2024-01-16 072843](https://github.com/Chlumsky/msdfgen/assets/86776302/b9dc52f9-7484-464e-aeca-e46b5bee255d)

They end up, having soft edges. In other words, they are anti-aliased to fill the between-pixel values. I don't want that

![Screenshot 2024-01-16 073109](https://github.com/Chlumsky/msdfgen/assets/86776302/dce7046f-0244-49e5-b8f0-c952d74e3c45)

But if I change the shader to a step function, the pixels are either on or off, leaving sharp but aliased edged

![Screenshot 2024-01-16 073119](https://github.com/Chlumsky/msdfgen/assets/86776302/dafb6689-5050-4d66-8d67-c8dfc7f533f1)

For regular fonts, I want anti-aliasing, as it just looks smoother, for pixelated fonts it just looks wrong.

The cleanest way would be to adjust the positioning by the fonts pixel grid offset, but as of now, I can't obtain that information and thanks to the offset being the same through all the shaded pixels, a simple step will fix that problem.

**Reply by Chlumsky** [reply]
Alright, so how I understand it now is that what I suggested at the beginning _is_ what you wanted for "pixelated fonts" but instead of a separate shader you wanted a parameter that would smoothly go between that and the regular version.

In that case, let's say we have a new shader variable `aaSmoothing` which at 1.0 is equivalent to the standard shader, at 0.0 it is pixelated, and other values make the anti-aliased border thicker (smoother) or narrower (coarser). Then, all you need to do is divide `pxRange` by `aaSmoothing`. However, as you can see, you will get a division by zero if you set it to zero, because that case is a singularity. Therefore, it might still be safer to have a condition `if (aaSmoothing == 0.0)` and a separate branch for that case (containing my earlier suggestion), although the shader will likely deal with +/-INF correctly.
